### PR TITLE
router: deprecate optional http filters and add route/vh level `is_optional` support in the typed_per_filter_config

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: router: remove optional http filters because it's hcm dependent
Additional Description:

The `ConfigImpl` (route table) created by the rds may shared by multiple listeners/hcm. It means that the `ConfigImpl` should be  hcm independent.

However, in the current codebase, the optional http filters is extracted from the HCM and used for the `ConfigImpl` creation.
Because the xds cache, the first hcm's http filters list will be used.

This typically a bug or potential error becasue different HCMs may have complete different http filters list and optional configuration.

Risk Level: mid.
Testing: n/a.
Docs Changes: n/a.
Release Notes: n/a.
Platform Specific Features: n/a.
[Optional Runtime guard:] n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
cc @markdroth because you concern #15025

**Comment by wbpcode** [comment]
There is a quesion about the PR self:

- should a runtime flag to be added? although i think this is a bug. But it was introduced at two years ago. May be some users depend on current behavior?  cc @envoyproxy/runtime-guard-changes cc @mattklein123

**Comment by soulxu** [comment]
> There is a quesion about the PR self:
> 
> * should a runtime flag to be added? although i think this is a bug. But it was introduced at two years ago. May be some users depend on current behavior?  cc @envoyproxy/runtime-guard-changes cc @mattklein123

As my understanding the usecase of `is_optional` is primary for the case of update, then new optional filter won't break the old envoy. I guess for this case, this new filter will be marked as optional at all HCMs' config, it doesn't make sense in some HCMs are optional and some HCMs aren't optional. I guess this is why this bug won't be found until now. If we directly remove this feature, I guess this will break some users depending on this.

I quickly search the code. maybe we can add the optional config into the hash of the provider 
https://github.com/envoyproxy/envoy/blob/7657d0b252e698e04d836c4481e71503d792e9c1/source/common/rds/route_config_provider_manager.cc#L82
Then we can get different provider instances for different optional config

Not sure whether this can be an acceptable temporary fix. if yes, then you can have a better fix later and won't take the risk break some users.

**Comment by wbpcode** [comment]
cc @soulxu See https://github.com/envoyproxy/envoy/pull/26097 and related discussion. I think it's similar to your new idea.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Please add some kind of test, thank you.

/wait

**Comment by wbpcode** [comment]
@mattklein123 I think I need some more input from the community first.

Will the suggestion from @soulxu be a possible better way to solve this problem? 
This is related to rds, so I want consider it more carefully and take various options into account.

**Comment by markdroth** [comment]
I think this is the right behavior, but I suggest combining this PR with #27210.  That way, any user that runs into a problem will be able to solve it with the `is_optional` mechanism.

I don't entirely understand @soulxu's suggestion, but I think the behavior we want here is:
- When we receive the RDS resource, we iterate through the filter configs, checking the type_url for each one to see if that filter impl is supported.
- If the filter impl is *not* supported, the default behavior is to NACK the RDS resource.
- If the filter impl is not supported but the entry is marked as `is_optional`, then we ignore the entry instead of NACKing.

Note that this behavior has already been agreed upon and has been implemented by gRPC.  I don't think we want Envoy to do something different.

**Comment by soulxu** [comment]
> * When we receive the RDS resource, we iterate through the filter configs, checking the type_url for each one to see if that filter impl is supported.
> * If the filter impl is _not_ supported, the default behavior is to NACK the RDS resource.
> * If the filter impl is not supported but the entry is marked as `is_optional`, then we ignore the entry instead of NACKing.
> 
Do we want the HCM's `is_optional` apply to the pre-route/vh filter config when `is_optional` isn't specified in the pre-route/vh filter config?

**Comment by wbpcode** [comment]
> I don't entirely understand @soulxu's suggestion, but I think the behavior we want here is:

It basically like #26097. We we shared same rds resource and proto config in multiple HCMs. But we may have different config providers based on the optional filters list. Rds resource will be accepted if all config providers validated it. 

But this way are more complicate and not sure if has other potential problem.

**Comment by soulxu** [comment]
> Of course we can fix current error directly like this PR did. But this error was exist for two years. I worry about the back compatiability. If some users depend on this 'error', orz.
> 
> Not sure if a runtime guard enough.

yea, the trouble is we documented this behavior in API doc https://github.com/envoyproxy/envoy/blob/bedd2b6cc19958eb5c57715722ccaf55b1fc7f87/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L1143-L1144

That is declared behavior, so in theory, removing it isn't right. except we consider this behavior is fully not working, then it is fine to delete it directly. But it work some case, not sure whether have people depend on it.

**Comment by wbpcode** [comment]
> I don't think that will work, because the set of HCM configs can change over time. Consider the case where we initially have 3 HCM configs that point to the same RDS resource, we ACK that RDS resource, and then later we get a new HCM config that does not work with the RDS resource. At that point, the RDS resource would suddenly be considered invalid, but it's too late to NACK it.

@markdroth yeah, I think you are right. I didn't thought this point in the past. Thanks.

**Comment by wbpcode** [comment]
Then, I think there no other choice. I will add a runtime guard and update doc. And also the test.

**Comment by wbpcode** [comment]
cc @alyssawilk for runtime guard.

**Comment by wbpcode** [comment]
@mattklein123 ready for a formal review

**Comment by alyssawilk** [comment]
re runtime guard, f it's high risk, or if it's a behavioral change (even if it's a bug fix) please runtime guard.

File: changelogs/current.yaml
Comment (alyssawilk) [inline changelogs/current.yaml:74]: sorry I'm confused what motivated this change.
I would assume (though API shepherds can correct me) that the optional tag was so one could use the same config for Envoy and gRPC xDS endpoints, where gRPC might not have certain filters implemented or vice versa.  don't we then want is_optional to err on the side of not validating rater than validating?
either way I'm going to ask Adi do primary review as I think it's a question of API and implementation

Comment (markdroth) [reply changelogs/current.yaml:74]: This PR is definitely the right thing from an xDS API perspective.

The purpose of `is_optional` is to allow control planes to configure new filters that may not yet be supported by all existing clients, regardless of whether those clients are Envoy or gRPC or anything else.  The `is_optional` field exists both in the HCM config for the top-level filter configs and also in the `typed_per_filter_config` overrides that exist at the VH, route, and cluster-weight levels.

Envoy's current behavior is that it uses the `is_optional` field at the HCM level when validating the RDS resource, rather than looking at the `is_optional` field in the RDS resource itself.  That behavior is totally broken with regard to resource cacheability, as per discussion in #26097.  This PR fixes Envoy to stop using the `is_optional` field at the HCM level to mark a filter as optional in `typed_per_filter_config` in the RDS resource.

Envoy currently does not use the `is_optional` field in `typed_per_filter_config` at all.  That problem is being fixed in #27210.

My only concern here is that we probably shouldn't leave users without any way to mark a filter as optional in `typed_per_filter_config`.  I think that means that we either need to merge #27210 before this PR, or we need to combine the two into a single PR.

Comment (wbpcode) [reply changelogs/current.yaml:74]: I think it would be more easy to review (and to cherry for third party folks) to keep these change in two different PR.
So, if no special requirement, I prefer to land #27210 first.

Comment (wbpcode) [reply changelogs/current.yaml:74]: Maybe I can split the #27210 to two pr, one for `is_optional` flag, one for `disabled` flag. The former one is simple and harmless. The latter one will bring some performance overhead and may need more consideration.
I will do it.

Comment (markdroth) [reply changelogs/current.yaml:74]: Good idea!  The `disabled` part of #27210 is not related to this PR, only the `is_optional` bit is, so splitting them up makes sense.

Once you've done that, I think there are the same two choices here: You can either include the `is_optional` field in this PR, or you can put that in a separate PR that gets landed before this one does.  Either one is fine with me.

Thanks!

Comment (wbpcode) [reply changelogs/current.yaml:74]: I will put the splited one into this PR. Then all things about the optional extension will be aggregated to this PR.

**Comment by wbpcode** [comment]
cc @adisuissa @markdroth for API review, thanks.

cc @snowp for code review, thanks.

File: source/common/router/config_impl.cc
Comment (snowp) [inline source/common/router/config_impl.cc:2154]: I think full_name() might already get you a static string? Maybe make this `std::string&` or `absl::string_view` to avoid the extra copy

File: source/common/router/config_impl.cc
Comment (snowp) [inline source/common/router/config_impl.cc:2151]: Since this is only used within the ctor I think we can avoid making it a class member?

File: source/common/router/config_impl.cc
Comment (snowp) [inline source/common/router/config_impl.cc:2163]: Do you think you could elaborate on what's going on here a bit more? It's bit hard to follow the flow and the motivation for the different flows

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, a few comments

**Comment by markdroth** [comment]
Please also update the PR title and description to reflect the addition of the `is_optional` field in `typed_per_filter_config`.  Thanks!

**Review by adisuissa — COMMENTED** [review]
Thanks for working on this!
Overall LGTM, would just clarify a few points.

It seems to me that the API allows overriding a config of some filter-type-A with a different filter-type-B, which is why the HCM's `is_optional` and the FilterConfig's `is_optional` may have different values. I think this use-case is only implied here, and maybe explaining that in the comments will help.

**Comment by markdroth** [comment]
> It seems to me that the API allows overriding a config of some filter-type-A with a different filter-type-B, which is why the HCM's `is_optional` and the FilterConfig's `is_optional` may have different values. I think this use-case is only implied here, and maybe explaining that in the comments will help.

No, that's not at all the intent.  It's not okay to have the same filter instance name with configs for two different filter types, and `is_optional` is not intended to change that.

Consider the following example:

```
// In HCM config, in LDS resource:
http_filters {
  name: "foo"
  typed_config {
    [types.googleapis.com/envoy.extensions.filters.http.rbac.v3.RBAC] {
      // ...
    }
  }
}

// In RDS resource:
typed_per_filter_config {
  key: "foo"
  value {
    [types.googleapis.com/envoy.extensions.filters.http.fault.v3.HTTPFault] {
      // ...
    }
  }
}
```

Note that both of these resources are valid and will be ACKed, because the client supports both the RBAC and HTTPFault filters.  The config is invalid only when the client attempts to put those two resources together, because the "foo" filter is actually an RBAC filter, and the RBAC filter doesn't understand the HTTPFault config.  The data plane needs to decide how to handle that: it can either fail requests at L7, or it can just ignore `typed_per_filter_config` entries whose configs are of the wrong type.

Note that we cannot NACK the RDS resource in this case, because when we're validating and caching the RDS resource, we have no way of knowing which set of HCM instances that RDS resource may be used in -- and that set may change over time, even after it's already too late to NACK the RDS resource.  The resource cache in the xDS transport protocol layer requires that each resource be validated independently of any other, which means that you cannot send a NACK based on a problem that is visible only when looking at a combination of more than one resource.

The `is_optional` field does not change any of the above.  The only purpose of `is_optional` is to allow new filter types to be configured by the control plane without breaking older clients that don't yet support them.  Its effect is to tell the client that if the filter type is unknown, then it can ignore the filter instead of NACKing the resource.  So, in the above example, if the `is_optional` field was set in both the LDS and RDS resource, the behavior would be:
- If the client supports the RBAC filter, the `is_optional` field in the LDS resource is a no-op.
- If the client does not support the RBAC filter, the filter entry would be ignored instead of causing the LDS resource to be NACKed.
- If the client supports the HTTPFault filter, the `is_optional` field in the RDS resource is a no-op.
- If the client does not support the HTTPFault filter, the filter entry would be ignored instead of causing the RDS resource to be NACKed.
- If the client supports the RBAC filter but not the HTTPFault filter, then it will configure the RBAC filter in the LDS resource, but it will not use the HTTPFault entry in the RDS resource, so there will be no conflict when putting the two resources together.
- If the client supports the HTTPFault filter but does not support the RBAC filter, then it will ignore the RBAC filter entry in the LDS resource, so there will be no filter instance with the name "foo".  It will see the HTTPFault filter in the RDS resource, but that will be a no-op, because there will not be a top-level filter under the name "foo" that could use that override config.

File: test/common/router/config_impl_test.cc
Comment (wbpcode) [reply test/common/router/config_impl_test.cc:10444]: sgtm, i will update it.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2154]: I am not sure the lifetime of full name that returned by the descriptor. So I create a static copy here to avoid possible illegal memory access.

I will check it and if the `full_name()` return static string, I will change it to string_view. :)

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2151]: good point, right.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2163]: I will add more comment about this feature, thanks.

File: source/common/router/config_impl.h
Comment (wbpcode) [reply source/common/router/config_impl.h:98]: right, thanks.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2184]: 🤔 , yeah, you suggesion is better.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2172]: > My assumption is that adding a FilterConfig can be used to change some non-specific-filter config (like the disabled field), and that's why it is ok for the code to ignore the entire config.

Yeah, we will implement the `disabled` field in another PR, so, here we just ignore the entire config first.

File: api/envoy/config/route/v3/route_components.proto
Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:2378]: I think it just add some more non-filter-specific flags/control options to custom the filter's behavior in specific route.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2175]: hmmm, is there some understand gaps? 

The `is_optional` flag in the FilterConfig of route is not used to override the `is_optional` flag in the HCM. They are two completely independent flags.

The one in the HCM will tell envoy a HTTP filter is optional when envoy loading the HCM. The one in the FilterConfig of route will tell envoy a HTTP filter ('s route configuration) is optional when loading the RouteConfiguration. 

I know in some way they are duplicate. But we have no better choice. We know nothing about the HCM's flags when we loading the RouteConfiguration, so additional/duplicated `is_optional` flag is necessary.

**Comment by wbpcode** [comment]
friendly ping @snowp @adisuissa

File: source/common/router/config_impl.cc
Comment (adisuissa) [reply source/common/router/config_impl.cc:2182]: I understand that `disabled` will be added in the future, but I think we are discussing what should happen if `FilterConfig.config` is empty.

At the moment if `FilterConfig.config` is empty, the filter will be skipped. In the future the behavior might be different. An example scenario: The HCM contains a filter that is disabled, and some Routes want to enable it.
If the expectation is that the `FilterConfig.config` could be empty, and the `FilterConfig.disabled` flag will enable it, then the current behavior will break previous configs, no?

File: source/common/router/config_impl.cc
Comment (adisuissa) [reply source/common/router/config_impl.cc:2185]: I think my issue is that AIUI, the `FilterConfig.config` can be of a filter type that is completely different than the HCM filter type. I understand this should not be encouraged, but I think this is possible.
Then the question is what should happen if the HCM filter is valid, but the `FilterConfig.config` points to a config that is unknown and optional. I think Mark's [comment](https://github.com/envoyproxy/envoy/pull/27263#issuecomment-1555096822) mentions this case:
>  * If the client supports the RBAC filter but not the HTTPFault filter, then it will configure the RBAC filter in the LDS resource, but it will not use the HTTPFault entry in the RDS resource, so there will be no conflict when putting the two resources together.

It is a bit unclear whether the RBAC filter should be used on the data-path when the route is used or not, and I just think the `FilterConfig` comments could be improved in this case.

**Review by adisuissa — COMMENTED** [review]
Thanks!
I tried clarifying some of my comments. LMK what you think.

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2182]: I don't think it's the case that an empty top-level filter config skips the filter.  The filter itself may be a no-op in that case, but I believe it still gets invoked.

Comment (wbpcode) [reply source/common/router/config_impl.cc:2182]: If the `FilterConfig.config` is empty, it just means the route level per filter config is nullptr and this route level per filter config will be ignored. But the filter will be enabled as normally. The `FilterConfig` is used to provide route level per filter config, it will not make the filter be skipped.

So, I don't think we will break the behavior.

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2185]: Forget `is_optional` for a moment.  How would you want to handle this if the data plane supports both filters?  I think we have basically two choices:

1. Fail requests at L7.
2. Ignore the `typed_per_filter_config` entry that's for the wrong filter type and use only the top-level config.

If we think that the answer in this case is (2), then `is_optional` doesn't matter, because we'd have exactly the same behavior regardless of whether the filter in the `typed_per_filter_config` entry was supported.

If we think that the answer is (1), then that might be an argument in favor of doing the same thing in the `is_optional` case.  But we'd have to actually do some additional bookkeeping to make this work: when we validate `typed_per_filter_config`, if we see an unsupported filter that's marked as `is_optional`, instead of skipping it completely, we'd need to record that the entry as an invalid config, so that when we put the LDS and RDS resources together, the RBAC filter can tell that the override config is invalid and can fail requests.

Personally, I think that the right answer for the case where both filters are supported is actually (1), which is what gRPC does.  However, for the case with an unsupported filter marked as `is_optional`, I don't think that it's worth the additional bookkeeping to make sure we have the same behavior; I'm fine with allowing that case to just ignore the override config.

Comment (wbpcode) [reply source/common/router/config_impl.cc:2185]: > I think my issue is that AIUI, the FilterConfig.config can be of a filter type that is completely different than the HCM filter type. I understand this should not be encouraged, but I think this is possible.

Yes. And it almost alyways different. Like `envoy.extensions.filters.http.ext_authz.v3.ExtAuthz` and `envoy.extensions.filters.http.ext_authz.v3.ExtAuthzPerRoute`, `envoy.extensions.filters.http.lua.v3.Lua` and `envoy.extensions.filters.http.lua.v3.LuaPerRoute`.

> Then the question is what should happen if the HCM filter is valid, but the FilterConfig.config points to a config that is unknown and optional.

The route level per filter config will be nullptr, but the filter will be executed.

File: api/envoy/config/route/v3/route_components.proto
Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:2378]: > It could be either way. It really depends on how the filter is implemented. For some filters, the semantic is that the per-route/vh config completely overrides the top-level config; for other filters, the two configs are merged together in some appropriate way.

**From perspective of external filter manager, the route level per filter config just a supplement to the filter config in the HCM**. But in some filters' implementations, the `supplement` may be used by priority.

File: api/envoy/config/route/v3/route_components.proto
Comment (soulxu) [inline api/envoy/config/route/v3/route_components.proto:2397]: It isn't clear about when the disabled is true and is_optional is false, will envoy reject the config if the filter doesn't support.

Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:2397]: > If true, the filter is disabled in the route or virtual host and the `config` field is ignored.

I think the comment has clarified this scenario. The `config` will be ignored. So no any validation will be done and the config will not be rejected.

File: source/common/router/config_impl.cc
Comment (soulxu) [reply source/common/router/config_impl.cc:2182]: Is there any meaningful usecase for FilterConfig without config specified? If not, we should mark the `config` as required.

**Review by adisuissa — DISMISSED** [review]
Thanks!
Overall LGTM, just one small thing regarding the tests that override the runtime guard.

/assign-from @envoyproxy/senior-maintainers

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2185]: See my comment above about `disabled`.  But I think that's mostly a separate issue; that problem already exists, since gRPC does have behavior (1) here.

For the current PR, I would argue that Envoy should also have behavior (1).  That way, when we eventually build a mechanism for clients to report the kind of config failures that are visible only when combining multiple resources, this will be considered an error and can be reported accordingly.

File: test/integration/scoped_rds_integration_test.cc
Comment (wbpcode) [reply test/integration/scoped_rds_integration_test.cc:468]: It should be removed. It would be a big clean-up.

Comment (wbpcode) [reply test/integration/scoped_rds_integration_test.cc:468]: I will add some comments here.

File: source/common/router/config_impl.cc
Comment (soulxu) [reply source/common/router/config_impl.cc:2182]: > It seems to me that an empty-config is a desired property, as it will allow (in the future) to set the filter as disabled for a specific route for example, w/o resending some filter config.

emm....if the config is empty, then there is no way to know which filter you are disabling.

Comment (wbpcode) [reply source/common/router/config_impl.cc:2182]: > I assume that NACKing was the behavior in Envoy prior to that PR (?), and it's definitely the behavior in gRPC. I don't think we should change that.

It's ok for me to NACK the empty config. Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly.
It's a break API change, but both envoy and grpc not use it for now, so it should be ok?

WDYT, @markdroth ? cc @adisuissa

Comment (adisuissa) [reply source/common/router/config_impl.cc:2182]: > It's ok for me to NACK the empty config. Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly. It's a break API change, but both envoy and grpc not use the `disabled` field for now, so it should be ok?

The change you describe does not break backwards compatibility, as the client will now accept a config which it previously rejected (this is similar to a field that has some valid range, and a PR increases the range - any config sent with the small range will also work with the wider range). Going the other way (decreasing the range) is not allowed.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2185]: Agree this should be a independent issue. I will mark this resolved here to make the context of this PR more clean.

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2182]: > > I assume that NACKing was the behavior in Envoy prior to that PR (?), and it's definitely the behavior in gRPC. I don't think we should change that.
> 
> I think envoy has no implementation for `FilterConfig` before this PR? My initial idea is that empty `config` + `disabled` could be used to express a filter is diabled explicitly or not.

Right, but the equivalent case can still happen without `FilterConfig`: the value in the `typed_per_filter_config` map can be empty.  I assume Envoy would NACK in that case?

> It's ok for me to NACK the empty config.

Okay, then let's do that.

> Then we need change the `disabled` to a wrappered bool. Then we can NACK the config only when the empty `config` is provided and `disabled` flag is not set explicitly. It's a break API change, but both envoy and grpc not use the `disabled` field for now, so it should be ok?

I don't quite understand what you mean by "wrapped bool".  Can you write exactly what the proto definition would look like?

I'm also not sure why the `config` field needs to be empty in the first place.  Presumably, it could still contain a config of the appropriate type, but it would simply not be used if `disabled` is set.  So control planes could continue to set it, just to make sure they don't break older clients that don't support the `disabled` field.

In any case, I don't think any of this remaining discussion actually needs to block this PR.  Let's move further discussion of this to #27210.

**Comment by alyssawilk** [comment]
I can do second pass but first can @adisuissa and @markdroth make sure to tag resolved conversation as resolved?  It's a bit hard to read at least parts of diffs with all the comments, and I don't want to resolve anything still in progress

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2182]: I don't think this issue is resolved.  The code still needs to be changed to NACK if the `config` field is unset.

Other than this, though, I have no other concerns with this PR.

**Comment by markdroth** [comment]
I don't have permission to resolve or unresolve comment threads.  But I commented on the one thread that @alyssawilk resolved that I think is still open.

File: source/common/router/config_impl.cc
Comment (wbpcode) [reply source/common/router/config_impl.cc:2182]: I will update this PR this weekend.

Comment (wbpcode) [reply source/common/router/config_impl.cc]: I think this is be resolved now.

**Comment by wbpcode** [comment]
Hi, @markdroth @adisuissa, could you give a stamp about this PR if it ready for a second pass? Then we can push this to the next phase. 😄

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo one last question about API breaking changes

File: source/common/router/config_impl.cc
Comment (markdroth) [reply source/common/router/config_impl.cc:2182]: This is not a breaking change.  This particular field was not used at all prior to this PR, because Envoy wasn't using the `FilterConfig` message at all.  And as @wbpcode pointed out, it wasn't previously allowed for the value in the `typed_per_filter_config` map to be unset, so this is consistent with the existing behavior there.

This looks good to me.

**Review by alyssawilk — DISMISSED** [review]

**Comment by wbpcode** [comment]
Tests were fixed and a new test to cover the new exception thowing was added.

Could you giva a new stamp? cc @alyssawilk Thanks.

**Review by alyssawilk — APPROVED** [review]

**Review by alyssawilk — APPROVED** [review]

**Comment by wbpcode** [comment]
close #15025

**Pull request merged**
